File: iosApp/LiveChatIOS/FirebaseMessagesBridge.swift
import Foundation
import FirebaseFirestore
import LiveChatCompose

final class FirebaseMessagesBridge: NSObject, MessagesRemoteBridge {
    private let db: Firestore
    private let config: FirebaseRestConfig
    private var listeners: [String: ListenerRegistration] = [:]

    init(config: FirebaseRestConfig) {
        self.db = Firestore.firestore()
        self.config = config
        super.init()
    }

    func startListening(recipientId: String, listener: MessagesRemoteListener) -> String {
        let token = UUID().uuidString
        let registration =
            messagesCollection(recipientId: recipientId)
                .addSnapshotListener { snapshot, error in
                    if let error = error {
                        listener.onError(message: error.localizedDescription)
                        return
                    }
                    let docs = snapshot?.documents ?? []
                    let payloads = docs.map { self.mapDocument($0) }
                    listener.onMessages(messages: payloads)
                }
        listeners[token] = registration
        return token
    }

    func stopListening(token: String) {
        listeners[token]?.remove()
        listeners.removeValue(forKey: token)
    }

    func fetchMessages(recipientId: String, completionHandler: @escaping ([TransportMessagePayload]?, Error?) -> Void) {
        messagesCollection(recipientId: recipientId).getDocuments { snapshot, error in
            if let error = error {
                completionHandler(nil, error)
                return
            }
            let docs = snapshot?.documents ?? []
            completionHandler(docs.map { self.mapDocument($0) }, nil)
        }
    }

    func sendMessage(
        recipientId: String,
        documentId: String,
        payload: TransportMessagePayload,
        completionHandler: @escaping (String?, Error?) -> Void
    ) {
        let document = messagesCollection(recipientId: recipientId).document(documentId)
        document.setData(payload.toFirestorePayload(), merge: false) { error in
            if let error = error {
                completionHandler(nil, error)
                return
            }
            completionHandler(document.documentID, nil)
        }
    }

    func deleteMessage(
        recipientId: String,
        documentId: String,
        completionHandler: @escaping @Sendable (Error?) -> Void
    ) {
        messagesCollection(recipientId: recipientId)
            .document(documentId)
            .delete { error in
                completionHandler(error)
            }
    }

    func ensureConversation(
        conversationId: String,
        completionHandler: @escaping @Sendable (Error?) -> Void
    ) {
        let doc = db.collection(config.conversationsCollection)
            .document(conversationId)
        doc.setData(["created_at": FieldValue.serverTimestamp()], merge: true) { error in
            completionHandler(error)
        }
    }


    private func messagesCollection(recipientId: String) -> CollectionReference {
        db.collection(config.conversationsCollection)
            .document(recipientId)
            .collection(config.messagesCollection)
    }

    private func mapDocument(_ doc: QueryDocumentSnapshot) -> TransportMessagePayload {
        let data = doc.data()
        let senderId = data["sender_id"] as? String
        let receiverId = data["receiver_id"] as? String
        let type = data["type"] as? String
        let content = data["content"] as? String
        let status = data["status"] as? String
        let createdAtMillis =
            (data["created_at_ms"] as? NSNumber)?.int64Value
            ?? (data["created_at"] as? Timestamp).map { Int64($0.dateValue().timeIntervalSince1970 * 1000) }

        let kotlinCreatedAt = createdAtMillis.map { KotlinLong(value: $0) }
        return TransportMessagePayload(
            id: doc.documentID,
            senderId: senderId,
            receiverId: receiverId,
            createdAtMillis: kotlinCreatedAt,
            type: type,
            content: content,
            status: status
        )
    }
}

private extension TransportMessagePayload {
    func toFirestorePayload() -> [String: Any] {
        var payload: [String: Any] = [
            "sender_id": senderId ?? "",
            "receiver_id": receiverId ?? "",
            "created_at": FieldValue.serverTimestamp(),
            "type": type ?? "text",
            "content": content ?? "",
            "status": status ?? "pending",
        ]
        if let createdAtMillis = createdAtMillis?.int64Value {
            payload["created_at_ms"] = createdAtMillis
        }
        return payload
    }
}

File: shared/data/src/androidMain/kotlin/com/edufelip/livechat/data/bridge/FirebaseMessagesBridge.kt
package com.edufelip.livechat.data.bridge

import com.edufelip.livechat.data.remote.FirebaseRestConfig
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.tasks.await
import java.util.UUID

class FirebaseMessagesBridge(
    private val firestore: FirebaseFirestore,
    private val config: FirebaseRestConfig,
) : MessagesRemoteBridge {
    private val listeners = mutableMapOf<String, ListenerRegistration>()

    override fun startListening(
        recipientId: String,
        listener: MessagesRemoteListener,
    ): String {
        val token = UUID.randomUUID().toString()
        val registration =
            inboundMessagesCollection(recipientId)
                .addSnapshotListener { snapshot, error ->
                    if (error != null) {
                        listener.onError(error.message ?: "Failed to listen for messages.")
                        return@addSnapshotListener
                    }
                    val payloads = snapshot?.documents?.mapNotNull { it.toPayloadOrNull() }.orEmpty()
                    listener.onMessages(payloads)
                }
        listeners[token] = registration
        return token
    }

    override fun stopListening(token: String) {
        listeners.remove(token)?.remove()
    }

    override suspend fun fetchMessages(recipientId: String): List<TransportMessagePayload> {
        val snapshot = inboundMessagesCollection(recipientId).get().await()
        return snapshot.documents.mapNotNull { it.toPayloadOrNull() }
    }

    override suspend fun sendMessage(
        recipientId: String,
        documentId: String,
        payload: TransportMessagePayload,
    ): String {
        val document = inboundMessagesCollection(recipientId).document(documentId)
        document.set(payload.toFirestorePayload()).await()
        return document.id
    }

    override suspend fun deleteMessage(
        recipientId: String,
        documentId: String,
    ) {
        inboundMessagesCollection(recipientId).document(documentId).delete().await()
    }

    override suspend fun ensureConversation(conversationId: String) {
        firestore.collection(config.conversationsCollection)
            .document(conversationId)
            .set(
                mapOf(
                    FIELD_OWNER_ID to conversationId,
                    FIELD_UPDATED_AT to FieldValue.serverTimestamp(),
                ),
                SetOptions.merge(),
            )
            .await()
    }

    private fun inboundMessagesCollection(recipientId: String) =
        firestore.collection(config.conversationsCollection)
            .document(recipientId)
            .collection(config.messagesCollection)

    private fun DocumentSnapshot.toPayloadOrNull(): TransportMessagePayload? =
        runCatching {
            val serverMillis = getTimestamp(FIELD_CREATED_AT)?.toDate()?.time
            TransportMessagePayload(
                id = id,
                senderId = getString(FIELD_SENDER_ID),
                createdAtMillis =
                    serverMillis
                        ?: getLong(FIELD_CREATED_AT_MS),
                createdAtServerMillis = serverMillis,
                content = getString(FIELD_CONTENT),
                actionType = getString(FIELD_ACTION_TYPE),
                messageId = getString(FIELD_MESSAGE_ID),
                actionMessageId = getString(FIELD_ACTION_MESSAGE_ID),
                contentType = getString(FIELD_CONTENT_TYPE),
            )
        }.getOrNull()

    private fun TransportMessagePayload.toFirestorePayload(): Map<String, Any?> =
        buildMap {
            put(FIELD_SENDER_ID, senderId)
            put(FIELD_CREATED_AT, FieldValue.serverTimestamp())
            put(FIELD_CREATED_AT_MS, createdAtMillis)
            val normalizedAction = actionType?.lowercase() ?: "message"
            put(FIELD_ACTION_TYPE, normalizedAction)
            if (normalizedAction == "message") {
                put(FIELD_MESSAGE_ID, messageId)
                put(FIELD_CONTENT, content)
                put(FIELD_CONTENT_TYPE, contentType)
            } else {
                put(FIELD_ACTION_MESSAGE_ID, actionMessageId)
            }
        }

    private companion object {
        const val FIELD_SENDER_ID = "sender_id"
        const val FIELD_CREATED_AT = "created_at"
        const val FIELD_CREATED_AT_MS = "created_at_ms"
        const val FIELD_CONTENT = "content"
        const val FIELD_ACTION_TYPE = "action_type"
        const val FIELD_MESSAGE_ID = "message_id"
        const val FIELD_ACTION_MESSAGE_ID = "action_message_id"
        const val FIELD_CONTENT_TYPE = "content_type"
        const val FIELD_OWNER_ID = "owner_id"
        const val FIELD_UPDATED_AT = "updated_at"
    }
}
